### ISSUE TITLE
Automate Tests

### ISSUE THREAD

**Issue by DanRStevens** [OutpostUniverse/OP2MapImager#41]
We could use some automated high level tests of the OP2MapImager program. These might run the compiled executable with specific command line parameters, and test for desired output, such as console text, or changes to the filesystem.

I believe such tests would be known as System Tests, or End-to-End Tests. I'm a little unclear on the difference between them. Seems like the distinction is more related to complex systems that have interacting executables, such as a main program, a database, and webservices.

[System Tests](http://softwaretestingfundamentals.com/system-testing/)
[End-to-End Tests](https://www.toolsqa.com/software-testing/what-does-end-to-end-test-mean/)

**Comment by DanRStevens** [comment]
@Brett: It sounds like you've been manually running the program with various inputs and checking for correct behavior. If you can document what commands you are running, and what output you are checking, that may provide a basis for writing such tests.

**Comment by DanRStevens** [comment]
I wanted to drop a link to this article from the Google Testing blog:
[Just Say No to More End-to-End Tests](https://testing.googleblog.com/2015/04/just-say-no-to-more-end-to-end-tests.html)

I read the article a few days ago and it is quite a good read. Despite the title, it does actually suggest having a few end-to-end tests, though suggests the bulk of tests should be unit tests. I would agree. A major reason being that unit tests help narrow down exactly where a bug is occurring, whereas end-to-end tests can take quite a while to diagnose and track down the cause of a failure.

**Comment by Brett208** [comment]
I read through the article 'Just Say No to More End-to-End Tests'. Overall, I agree with the pyramid approach where you have many more unit tests than end to end tests.

A couple of counterpoints to the thoughts in the article:

1. It mentions having to wait until the end of the day / long compile to run tests. OP2MapImager is a small program and the time difference between running unit tests vs end to end tests would be minimal. The point is more valid for a larger program.

2. The article mentions having many developers tracking the same bug down since the end to end test fails on a specific bug that hides others. Since at most two of us are working on the project in spare time, not really a problem since we really only need to work one problem at a time. Again, definitely more of a concern for a larger team.

3. The article mentions difficulty in tracing bugs in end-to-end tests because the bugs could be in large tracts of the program. Assuming we actually well covered OP2Utility with unit tests and assuming that FreeImage is fairly bugfree, the amount of code in OP2MapImager is fairly small. There isn't all that much code to sift through to find the bugs. Again, this is certainly a bigger concern on a much larger project.

We can unit test fair portions of the code that is contained in OP2Utility. A decent number of our path problems are stemming from the ResourceManager and XFile, which is all available for unit tests. We just hadn't implemented the tests yet. 

So, I guess I would be okay with implementing some End to End tests and skipping the unit tests. Not that I'm against adding both if someone wants to invest the time, just not sure I want to focus on it.

The bigger problem to me is that we don't have an actual map or saved game to upload into the repository. We could add some tests to see what happens if the file isn't found or is malformed/empty. Or we could host the maps somewhere else to complete the end to end tests if that is palatable.

-Brett

**Comment by DanRStevens** [comment]
I agree with your points. Maybe a little less on #3 though. Our projects aren't big, but unit tests certainly do allow for quicker determination of the root cause as opposed to end-to-end tests. There have been a few cases already where manual end-to-end testing has revealed a bug, but we were uncertain or disagreed on where the bug actually was. Granted, in one of those cases, we did have unit tests, they just weren't sufficient and missed an important case, which showed up in the manual end-to-end testing.

I agree about the pyramid structure for testing. And I would like to practice habits that scale well. Currently though, we don't have any end-to-end tests, so I'd say that's too few at the top of the pyramid. :)

----

In terms of using (possibly proprietary) map or saved game data to test with, that doesn't need to be a show stopper. We can still write the tests, and just not run them on the CI servers. It could be that we only run the end-to-end tests locally. Maybe we could add an "e2e" rule to the Makefile that would run end-to-end tests, which would be completely separate from the unit tests.

Come to think of it, I once saw Makefiles themselves being used for end-to-end testing. Each test became a make rule, with the recipe being the executable to run with the various parameters you were testing it with. It was way slower than unit tests though.

----

I'd say we should start on some kind of end-to-end test, just so we figure out how to actually write one. We should add more unit tests too. Though perhaps having some end-to-end tests would encourage us to increase our unit testing, like that cases where a manual e2e test revealed a bug in path handling that the unit tests missed.
